lib/services/resource_test.go

```diff
@@ -296,3 +299,53 @@ func TestProtoResourceRoundtrip(t *testing.T) {
 		})
 	}
 }
+
+func TestConvertResource(t *testing.T) {
+	healthCfg, err := healthcheckconfig.NewHealthCheckConfig("example",
+		&healthcheckconfigv1.HealthCheckConfigSpec{
+			Match: &healthcheckconfigv1.Matcher{
+				DbLabels: []*labelv1.Label{{
+					Name:   types.Wildcard,
+					Values: []string{types.Wildcard},
+				}},
+			},
+		},
+	)
+	require.NoError(t, err)
+
+	user := &types.UserV2{
+		Kind: "user",
+		Metadata: types.Metadata{
+			Name: "llama",
+		},
+		Spec: types.UserSpecV2{
+			Roles: []string{"human", "camelidae"},
+		},
+	}
+
+	t.Run("converts legacy to RFD 153 resource", func(t *testing.T) {
+		resource := types.Resource153ToLegacy(healthCfg)
+		got, err := convertResource[*healthcheckconfigv1.HealthCheckConfig](resource)
+		require.NoError(t, err)
+		require.Equal(t, healthCfg, got)
+	})
+
+	t.Run("converts legacy to legacy", func(t *testing.T) {
+		resource := user.DeepCopy()
+		got, err := convertResource[*types.UserV2](types.Resource(resource))
+		require.NoError(t, err)
+		require.Equal(t, user, got)
+	})
+
+	t.Run("handles unexpected RFD 153 resource", func(t *testing.T) {
+		resource := types.Resource153ToLegacy(healthCfg)
+		_, err := convertResource[*types.Server](resource)
+		require.ErrorContains(t, err, "expected resource type *types.Server, got *types.resource153ToLegacyAdapter")
```

Comment:
nit: would be nice to see the actual type instead of *types.resource153ToLegacyAdapter